lib/internal/modules/run_main.js

```diff
@@ -12,19 +12,33 @@ const path = require('path');
  * @param {string} main - Entry point path
  */
 function resolveMainPath(main) {
-  // Note extension resolution for the main entry point can be deprecated in a
-  // future major.
-  // Module._findPath is monkey-patchable here.
-  const { Module } = require('internal/modules/cjs/loader');
-  let mainPath = Module._findPath(path.resolve(main), null, true);
+  if (getOptionValue('--preserve-symlinks-main')) { return; }
+
+  const defaultType = getOptionValue('--experimental-default-type');
+  /** @type {string} */
+  let mainPath;
+  if (defaultType === 'module') {
+    mainPath = path.resolve(main);
+  } else {
+    // Extension searching for the main entry point is supported only in legacy mode.
+    // Module._findPath is monkey-patchable here.
+    const { Module } = require('internal/modules/cjs/loader');
+    mainPath = Module._findPath(path.resolve(main), null, true);
+  }
   if (!mainPath) { return; }
 
-  const preserveSymlinksMain = getOptionValue('--preserve-symlinks-main');
-  if (!preserveSymlinksMain) {
-    const { toRealPath } = require('internal/modules/helpers');
+  // Follow symlinks
+  const { toRealPath } = require('internal/modules/helpers');
+  try {
     mainPath = toRealPath(mainPath);
+  } catch (err) {
+    if (err.code === 'ENOENT' && defaultType === 'module') {
```

Comment:
```suggestion
    if (defaultType === 'module' && err?.code === 'ENOENT') {
```

Reply:
How could `err` be undefined?

Reply:
Why would it not be? It's important to protect the error path from this kind of stuff, because not doing it does not bring us anything, and increases the risk for worsening the error message.